common/src/tracer.rs

```diff
@@ -1,14 +1,101 @@
+/// indicate the method's first parameter is a bool, however here the first parameter is an Option<&str>
+/// with the URL of the OpenTelemetry collector to be used, or None, if OpenTelemetry is not used.
+#[macro_export]
+macro_rules! conditional_init_tracer {
+    ($otel_endpoint:expr, $service_name:literal) => {
+        init_tracer!($service_name, $otel_endpoint, LevelFilter::TRACE)
+    };
+    ($otel_endpoint:expr, $service_name:literal, $level: expr) => {
+        $otel_endpoint
+            .map(|otel_endpoint| {
+                OtelTracer::new(otel_endpoint, $service_name, Some((module_path!(), $level)))
+                    .expect("Open Telemetry Tracer is created")
+            })
+            .or_else(|| {
+                tracing_subscriber::fmt::init();
+                None
+            })
+    };
+}
+
+/// May be used when the component produces messages.
+/// The `conditional_` prefix indicates a bool should be passed,
+/// indicating whether OpenTelemetry is used.
+/// If this is false, the methods usually do nothing.
+pub trait FutureRecordTracerExt {
+    fn optional_headers(self, headers: Option<OwnedHeaders>) -> Self;
+    fn conditional_inject_current_span_into_headers(self, use_otel: bool) -> Self;
+    fn conditional_inject_span_into_headers(self, use_otel: bool, span: &Span) -> Self;
+}
+
+impl FutureRecordTracerExt for FutureRecord<'_, str, [u8]> {
+    fn optional_headers(self, headers: Option<OwnedHeaders>) -> Self {
+        if let Some(headers) = headers {
+            self.headers(headers)
+        } else {
+            self
+        }
+    }
+
+    fn conditional_inject_current_span_into_headers(self, use_otel: bool) -> Self {
+        self.conditional_inject_span_into_headers(use_otel, &tracing::Span::current())
+    }
+
+    fn conditional_inject_span_into_headers(self, use_otel: bool, span: &Span) -> Self {
+        if use_otel {
+            let mut headers = OwnedHeaders::new();
+            opentelemetry::global::get_text_map_propagator(|propagator| {
+                propagator.inject_context(&span.context(), &mut HeaderInjector(&mut headers))
+            });
+            self.headers(headers)
+        } else {
+            self
+        }
+    }
+}
+
+/// May be used when the component consumne messages.
+/// The `conditional_` prefix indicates a bool should be passed,
+/// indicating whether OpenTelemetry is used.
+/// If this is false, the methods usually do nothing.
+pub trait OptionalHeaderTracerExt {
+    fn conditional_extract_to_current_span(self, use_otel: bool);
+    fn conditional_extract_to_span(self, use_otel: bool, span: &Span);
+}
+
+impl OptionalHeaderTracerExt for Option<&BorrowedHeaders> {
+    fn conditional_extract_to_current_span(self, use_otel: bool) {
+        self.conditional_extract_to_span(use_otel, &tracing::Span::current())
+    }
+
+    fn conditional_extract_to_span(self, use_otel: bool, span: &Span) {
```

Comment:
What does this do if `use_otel` is `true`, the message has headers but not headers that were added by OpenTelemetry?

Reply:
If it can't find valid OpenTelemetry headers, then it sets the parent of `span` to the current span, effectively it does nothing in this case.

Reply:
I will make a modification however, so that it adds to existing headers, rather than overwriting them. However it must push them to the front of the headers as this is where OpenTelemetry reads them from.